### PR TITLE
Unroll StringBuilder.Append for const string

### PR DESCRIPTION
**Pull request by yesmey**
Buffer.Memmove is now being unrolled for constant lengths. If we simplify Append(ref char, int) a bit, the JIT can inline it.
As a result, this allows the methods StringBuilder.Append(string) and StringBuilder.AppendLine() to be able to be unrolled.
I didn't seem to need to add any AggressiveInlining attribute, so I left it out.

**Const string example**

```csharp
public static void Example(StringBuilder stringBuilder)
{
    stringBuilder.Append("1234567890abcdefghijklmnopqrstuvwxyzåäö");
}
```

Before:
```assembly
; Method Program:Example(System.Text.StringBuilder)
G_M35345_IG01:  ;; offset=0000H
       sub      rsp, 40
						;; size=4 bbWeight=1 PerfScore 0.25

G_M35345_IG02:  ;; offset=0004H
       cmp      byte  ptr [rcx], cl
       mov      rdx, 0x154802028D8      ; '1'
       add      rdx, 12
       mov      r8d, 39
       call     [System.Text.StringBuilder:Append(byref,int):this]
       nop      
						;; size=29 bbWeight=1 PerfScore 7.00

G_M35345_IG03:  ;; offset=0021H
       add      rsp, 40
       ret      
						;; size=5 bbWeight=1 PerfScore 1.25
; Total bytes of code: 38
```

After:

```assembly
; Method Program:Example(System.Text.StringBuilder)
G_M35345_IG01:  ;; offset=0000H
       sub      rsp, 40
       vzeroupper 
						;; size=7 bbWeight=1 PerfScore 1.25

G_M35345_IG02:  ;; offset=0007H
       mov      rdx, 0x18C802028D8      ; '1'
       add      rdx, 12
       mov      r8, gword ptr [rcx+08H]
       mov      eax, dword ptr [rcx+18H]
       lea      r9d, [rax+27H]
       cmp      dword ptr [r8+08H], r9d
       jb       SHORT G_M35345_IG04
						;; size=31 bbWeight=1 PerfScore 9.00

G_M35345_IG03:  ;; offset=0026H
       movsxd   r9, eax
       lea      r8, bword ptr [r8+2*r9+10H]
       vmovdqu  ymm0, ymmword ptr [rdx]
       vmovdqu  ymm1, ymmword ptr [rdx+20H]
       vmovdqu  xmm2, xmmword ptr [rdx+3EH]
       vmovdqu  ymmword ptr [r8], ymm0
       vmovdqu  ymmword ptr [r8+20H], ymm1
       vmovdqu  xmmword ptr [r8+3EH], xmm2
       add      eax, 39
       mov      dword ptr [rcx+18H], eax
       jmp      SHORT G_M35345_IG05
						;; size=47 bbWeight=0.50 PerfScore 12.25

G_M35345_IG04:  ;; offset=0055H
       mov      r8d, 39
       call     [System.Text.StringBuilder:AppendWithExpansion(byref,int):this]
						;; size=12 bbWeight=0.50 PerfScore 1.62

G_M35345_IG05:  ;; offset=0061H
       nop      
						;; size=1 bbWeight=1 PerfScore 0.25

G_M35345_IG06:  ;; offset=0062H
       add      rsp, 40
       ret      
						;; size=5 bbWeight=1 PerfScore 1.25
; Total bytes of code: 103
```

**AppendLine example**

```csharp
public static void Example(StringBuilder stringBuilder)
{
    stringBuilder.AppendLine();
}
```

Before:
```assembly
; Method Program:Example(System.Text.StringBuilder)
G_M35345_IG01:  ;; offset=0000H
       sub      rsp, 40
						;; size=4 bbWeight=1 PerfScore 0.25

G_M35345_IG02:  ;; offset=0004H
       cmp      byte  ptr [rcx], cl
       mov      rdx, 0x195802028D8      ; '  '
       add      rdx, 12
       mov      r8d, 2
       call     [System.Text.StringBuilder:Append(byref,int):this]
       nop      
						;; size=29 bbWeight=1 PerfScore 7.00

G_M35345_IG03:  ;; offset=0021H
       add      rsp, 40
       ret      
						;; size=5 bbWeight=1 PerfScore 1.25
; Total bytes of code: 38
```

After:
```assembly
; Method Program:Example(System.Text.StringBuilder)
G_M35345_IG01:  ;; offset=0000H
       sub      rsp, 40
						;; size=4 bbWeight=1 PerfScore 0.25

G_M35345_IG02:  ;; offset=0004H
       mov      rdx, 0x17C002028D8      ; '  '
       add      rdx, 12
       mov      r8, gword ptr [rcx+08H]
       mov      eax, dword ptr [rcx+18H]
       lea      r9d, [rax+02H]
       cmp      dword ptr [r8+08H], r9d
       jb       SHORT G_M35345_IG04
						;; size=31 bbWeight=1 PerfScore 9.00

G_M35345_IG03:  ;; offset=0023H
       movsxd   r9, eax
       lea      r8, bword ptr [r8+2*r9+10H]
       mov      r9d, dword ptr [rdx]
       mov      dword ptr [r8], r9d
       add      eax, 2
       mov      dword ptr [rcx+18H], eax
       jmp      SHORT G_M35345_IG05
						;; size=22 bbWeight=0.50 PerfScore 3.75

G_M35345_IG04:  ;; offset=0039H
       mov      r8d, 2
       call     [System.Text.StringBuilder:AppendWithExpansion(byref,int):this]
						;; size=12 bbWeight=0.50 PerfScore 1.62

G_M35345_IG05:  ;; offset=0045H
       nop      
						;; size=1 bbWeight=1 PerfScore 0.25

G_M35345_IG06:  ;; offset=0046H
       add      rsp, 40
       ret      
						;; size=5 bbWeight=1 PerfScore 1.25
; Total bytes of code: 75
```

@EgorBo can you please take a look at this

### CODE REVIEW

**Comment by EgorBo** [comment]
Very nice use-case! One problem that it might slightly regress the case when it's not inlined and Mono.
The method seems to be compiled in a small codegen (when not unrolled) so presumably it won't hurt to mark it as AggressiveInlining.

Another option is to tune inliner - if it sees `Buffer.Memmove` and the last arg is a constant - boost inlining (but you need to invoke a different overload of Buffer.Memmove, the one that accepts `byte`)

**Comment by EgorBo** [comment]
https://github.com/EgorBo/runtime-1/commit/b79af3ff356596d13083dce0fda0b0acd9a0af5a

this seems to be enough to make it inlineable for const input (it gives inliner extra hint)

**Comment by yesmey** [comment]
With current code:
```csharp
public static void Example(StringBuilder builder, string s)
{
    builder.Append(s);
    builder.Append("1234567890abcdefghijklmnopqrstuvwxyzABCDEFGHIJKLMN");
    builder.AppendLine();
}
```

```assembly
; Method Program:Example(System.Text.StringBuilder,System.String)
G_M17123_IG01:  ;; offset=0000H
       push     rsi
       sub      rsp, 32
       vzeroupper 
       mov      rsi, rcx
						;; size=11 bbWeight=1 PerfScore 2.50

G_M17123_IG02:  ;; offset=000BH
       cmp      byte  ptr [rsi], sil
       test     rdx, rdx
       je       SHORT G_M17123_IG04
						;; size=8 bbWeight=1 PerfScore 4.25

G_M17123_IG03:  ;; offset=0013H
       mov      r8d, dword ptr [rdx+08H]
       add      rdx, 12
       mov      rcx, rsi
       call     [System.Text.StringBuilder:Append(byref,int):this]
						;; size=17 bbWeight=0.50 PerfScore 2.75

G_M17123_IG04:  ;; offset=0024H
       mov      rdx, 0x1EB802028D8      ; '1234567890abcdefghijklmnopqrstuvwxyzABCDEFGHIJKLMN'
       add      rdx, 12
       mov      rcx, gword ptr [rsi+08H]
       mov      r8d, dword ptr [rsi+18H]
       lea      eax, [r8+32H]
       cmp      dword ptr [rcx+08H], eax
       jb       SHORT G_M17123_IG06
						;; size=31 bbWeight=1 PerfScore 9.00

G_M17123_IG05:  ;; offset=0043H
       movsxd   rax, r8d
       lea      rcx, bword ptr [rcx+2*rax+10H]
       vmovdqu  ymm0, ymmword ptr [rdx]
       vmovdqu  ymm1, ymmword ptr [rdx+20H]
       vmovdqu  ymm2, ymmword ptr [rdx+40H]
       vmovdqu  xmm3, xmmword ptr [rdx+54H]
       vmovdqu  ymmword ptr [rcx], ymm0
       vmovdqu  ymmword ptr [rcx+20H], ymm1
       vmovdqu  ymmword ptr [rcx+40H], ymm2
       vmovdqu  xmmword ptr [rcx+54H], xmm3
       add      r8d, 50
       mov      dword ptr [rsi+18H], r8d
       jmp      SHORT G_M17123_IG07
						;; size=56 bbWeight=0.50 PerfScore 15.75

G_M17123_IG06:  ;; offset=007BH
       mov      rcx, rsi
       mov      r8d, 50
       call     [System.Text.StringBuilder:AppendWithExpansion(byref,int):this]
						;; size=15 bbWeight=0.50 PerfScore 1.75

G_M17123_IG07:  ;; offset=008AH
       mov      rdx, 0x1EB80202958      ; '  '
       add      rdx, 12
       mov      rcx, gword ptr [rsi+08H]
       mov      r8d, dword ptr [rsi+18H]
       lea      eax, [r8+02H]
       cmp      dword ptr [rcx+08H], eax
       jb       SHORT G_M17123_IG09
						;; size=31 bbWeight=1 PerfScore 9.00

G_M17123_IG08:  ;; offset=00A9H
       movsxd   rdx, r8d
       lea      rcx, bword ptr [rcx+2*rdx+10H]
       mov      word  ptr [rcx], 13
       mov      word  ptr [rcx+02H], 10
       mov      dword ptr [rsi+18H], eax
       jmp      SHORT G_M17123_IG10
						;; size=24 bbWeight=0.50 PerfScore 3.12

G_M17123_IG09:  ;; offset=00C1H
       mov      rcx, rsi
       mov      r8d, 2
       call     [System.Text.StringBuilder:AppendWithExpansion(byref,int):this]
						;; size=15 bbWeight=0.50 PerfScore 1.75

G_M17123_IG10:  ;; offset=00D0H
       nop      
						;; size=1 bbWeight=1 PerfScore 0.25

G_M17123_IG11:  ;; offset=00D1H
       add      rsp, 32
       pop      rsi
       ret      
						;; size=6 bbWeight=1 PerfScore 1.75
; Total bytes of code: 215
```

**Comment by EgorBo** [comment]
@yesmey can you apply this patch https://github.com/dotnet/runtime/commit/47eb0eb853c0ae80e47eb556264bf2008e54a7c2

it allows us to avoid using uglier version of Memmove + possibly improves other similiar cases

**Comment by yesmey** [comment]
> @yesmey can you apply this patch [47eb0eb](https://github.com/dotnet/runtime/commit/47eb0eb853c0ae80e47eb556264bf2008e54a7c2)
> 
> it allows us to avoid using uglier version of Memmove + possibly improves other similiar cases

Works perfectly for this scenario, thanks! I'll post some examples here. I assume you will take it from here and close this?

<details>
<summary>Append non-const</summary>

```csharp
public static void Append(StringBuilder builder, string s)
{
    builder.Append(s);
}
```

***Before***
```assembly
; Method Program:Append(System.Text.StringBuilder,System.String)
G_M43621_IG01:  ;; offset=0000H
       sub      rsp, 40
						;; size=4 bbWeight=1 PerfScore 0.25

G_M43621_IG02:  ;; offset=0004H
       cmp      byte  ptr [rcx], cl
       test     rdx, rdx
       je       SHORT G_M43621_IG04
						;; size=7 bbWeight=1 PerfScore 4.25

G_M43621_IG03:  ;; offset=000BH
       mov      r8d, dword ptr [rdx+08H]
       add      rdx, 12
       call     [System.Text.StringBuilder:Append(byref,int):this]
						;; size=14 bbWeight=0.49 PerfScore 2.57

G_M43621_IG04:  ;; offset=0019H
       nop      
						;; size=1 bbWeight=1 PerfScore 0.25

G_M43621_IG05:  ;; offset=001AH
       add      rsp, 40
       ret      
						;; size=5 bbWeight=1 PerfScore 1.25
; Total bytes of code: 31
```

***After***
```assembly
; Method Program:Append(System.Text.StringBuilder,System.String)
G_M43621_IG01:  ;; offset=0000H
       sub      rsp, 40
						;; size=4 bbWeight=1 PerfScore 0.25

G_M43621_IG02:  ;; offset=0004H
       cmp      byte  ptr [rcx], cl
       test     rdx, rdx
       je       SHORT G_M43621_IG04
						;; size=7 bbWeight=1 PerfScore 4.25

G_M43621_IG03:  ;; offset=000BH
       mov      r8d, dword ptr [rdx+08H]
       add      rdx, 12
       call     [System.Text.StringBuilder:Append(byref,int):this]
						;; size=14 bbWeight=0.50 PerfScore 2.62

G_M43621_IG04:  ;; offset=0019H
       nop      
						;; size=1 bbWeight=1 PerfScore 0.25

G_M43621_IG05:  ;; offset=001AH
       add      rsp, 40
       ret      
						;; size=5 bbWeight=1 PerfScore 1.25
; Total bytes of code: 31
```

</details>

<details>
<summary>Append by index non-const</summary>

```csharp
public static void AppendIndex(StringBuilder builder, string s)
{
    builder.Append(s, 2, 16);
}
```

***Before***
```assembly
; Method Program:AppendIndex(System.Text.StringBuilder,System.String)
G_M19771_IG01:  ;; offset=0000H
       sub      rsp, 40
						;; size=4 bbWeight=1 PerfScore 0.25

G_M19771_IG02:  ;; offset=0004H
       mov      r8d, 2
       mov      r9d, 16
       cmp      dword ptr [rcx], ecx
       call     [System.Text.StringBuilder:Append(System.String,int,int):System.Text.StringBuilder:this]
       nop      
						;; size=21 bbWeight=1 PerfScore 6.75

G_M19771_IG03:  ;; offset=0019H
       add      rsp, 40
       ret      
						;; size=5 bbWeight=1 PerfScore 1.25
; Total bytes of code: 30
```

***After***
```assembly
; Method Program:AppendIndex(System.Text.StringBuilder,System.String)
G_M19771_IG01:  ;; offset=0000H
       sub      rsp, 40
						;; size=4 bbWeight=1 PerfScore 0.25

G_M19771_IG02:  ;; offset=0004H
       mov      r8d, 2
       mov      r9d, 16
       cmp      dword ptr [rcx], ecx
       call     [System.Text.StringBuilder:Append(System.String,int,int):System.Text.StringBuilder:this]
       nop      
						;; size=21 bbWeight=1 PerfScore 6.75

G_M19771_IG03:  ;; offset=0019H
       add      rsp, 40
       ret      
						;; size=5 bbWeight=1 PerfScore 1.25
; Total bytes of code: 30
```

</details>

<details>
<summary>AppendLine</summary>

```csharp
public static void AppendLine(StringBuilder builder, string s)
{
    builder.AppendLine(s);
}
```

***Before***
```assembly
; Method Program:AppendLine(System.Text.StringBuilder,System.String)
G_M48267_IG01:  ;; offset=0000H
       push     rsi
       sub      rsp, 32
       mov      rsi, rcx
						;; size=8 bbWeight=1 PerfScore 1.50

G_M48267_IG02:  ;; offset=0008H
       cmp      byte  ptr [rsi], sil
       test     rdx, rdx
       je       SHORT G_M48267_IG04
						;; size=8 bbWeight=1 PerfScore 4.25

G_M48267_IG03:  ;; offset=0010H
       mov      r8d, dword ptr [rdx+08H]
       add      rdx, 12
       mov      rcx, rsi
       call     [System.Text.StringBuilder:Append(byref,int):this]
						;; size=17 bbWeight=0.49 PerfScore 2.70

G_M48267_IG04:  ;; offset=0021H
       mov      rdx, 0x283002028D8      ; '  '
       add      rdx, 12
       mov      rcx, rsi
       mov      r8d, 2
       call     [System.Text.StringBuilder:Append(byref,int):this]
       nop      
						;; size=30 bbWeight=1 PerfScore 4.25

G_M48267_IG05:  ;; offset=003FH
       add      rsp, 32
       pop      rsi
       ret      
						;; size=6 bbWeight=1 PerfScore 1.75
; Total bytes of code: 69
```

***After***
```assembly
; Method Program:AppendLine(System.Text.StringBuilder,System.String)
G_M48267_IG01:  ;; offset=0000H
       push     rsi
       sub      rsp, 32
       mov      rsi, rcx
						;; size=8 bbWeight=1 PerfScore 1.50

G_M48267_IG02:  ;; offset=0008H
       cmp      byte  ptr [rsi], sil
       test     rdx, rdx
       je       SHORT G_M48267_IG04
						;; size=8 bbWeight=1 PerfScore 4.25

G_M48267_IG03:  ;; offset=0010H
       mov      r8d, dword ptr [rdx+08H]
       add      rdx, 12
       mov      rcx, rsi
       call     [System.Text.StringBuilder:Append(byref,int):this]
						;; size=17 bbWeight=0.50 PerfScore 2.75

G_M48267_IG04:  ;; offset=0021H
       mov      rdx, 0x207003028D8      ; '  '
       add      rdx, 12
       mov      rcx, gword ptr [rsi+08H]
       mov      r8d, dword ptr [rsi+18H]
       lea      eax, [r8+02H]
       cmp      dword ptr [rcx+08H], eax
       jb       SHORT G_M48267_IG06
						;; size=31 bbWeight=1 PerfScore 9.00

G_M48267_IG05:  ;; offset=0040H
       movsxd   rdx, r8d
       lea      rcx, bword ptr [rcx+2*rdx+10H]
       mov      word  ptr [rcx], 13
       mov      word  ptr [rcx+02H], 10
       mov      dword ptr [rsi+18H], eax
       jmp      SHORT G_M48267_IG07
						;; size=24 bbWeight=0.50 PerfScore 3.12

G_M48267_IG06:  ;; offset=0058H
       mov      rcx, rsi
       mov      r8d, 2
       call     [System.Text.StringBuilder:AppendWithExpansion(byref,int):this]
						;; size=15 bbWeight=0.50 PerfScore 1.75

G_M48267_IG07:  ;; offset=0067H
       nop      
						;; size=1 bbWeight=1 PerfScore 0.25

G_M48267_IG08:  ;; offset=0068H
       add      rsp, 32
       pop      rsi
       ret      
						;; size=6 bbWeight=1 PerfScore 1.75
; Total bytes of code: 110
```

</details>

<details>
<summary>Append const</summary>

```csharp
public static void AppendConst(StringBuilder builder)
{
    builder.Append("1234567890abcdefghijklmnopqrstuvwxyzABCDEFGHIJKLMN");
}
```

***Before***
```assembly
; Method Program:AppendConst(System.Text.StringBuilder)
G_M5650_IG01:  ;; offset=0000H
       sub      rsp, 40
						;; size=4 bbWeight=1 PerfScore 0.25

G_M5650_IG02:  ;; offset=0004H
       cmp      byte  ptr [rcx], cl
       mov      rdx, 0x14A003028D8      ; '1234567890abcdefghijklmnopqrstuvwxyzABCDEFGHIJKLMN'
       add      rdx, 12
       mov      r8d, 50
       call     [System.Text.StringBuilder:Append(byref,int):this]
       nop      
						;; size=29 bbWeight=1 PerfScore 7.00

G_M5650_IG03:  ;; offset=0021H
       add      rsp, 40
       ret      
						;; size=5 bbWeight=1 PerfScore 1.25
; Total bytes of code: 38
```

***After***
```assembly
; Method Program:AppendConst(System.Text.StringBuilder)
G_M5650_IG01:  ;; offset=0000H
       sub      rsp, 40
       vzeroupper 
						;; size=7 bbWeight=1 PerfScore 1.25

G_M5650_IG02:  ;; offset=0007H
       mov      rdx, 0x20B002028D8      ; '1234567890abcdefghijklmnopqrstuvwxyzABCDEFGHIJKLMN'
       add      rdx, 12
       mov      r8, gword ptr [rcx+08H]
       mov      eax, dword ptr [rcx+18H]
       lea      r9d, [rax+32H]
       cmp      dword ptr [r8+08H], r9d
       jb       SHORT G_M5650_IG04
						;; size=31 bbWeight=1 PerfScore 9.00

G_M5650_IG03:  ;; offset=0026H
       movsxd   r9, eax
       lea      r8, bword ptr [r8+2*r9+10H]
       vmovdqu  ymm0, ymmword ptr [rdx]
       vmovdqu  ymm1, ymmword ptr [rdx+20H]
       vmovdqu  ymm2, ymmword ptr [rdx+40H]
       vmovdqu  xmm3, xmmword ptr [rdx+54H]
       vmovdqu  ymmword ptr [r8], ymm0
       vmovdqu  ymmword ptr [r8+20H], ymm1
       vmovdqu  ymmword ptr [r8+40H], ymm2
       vmovdqu  xmmword ptr [r8+54H], xmm3
       add      eax, 50
       mov      dword ptr [rcx+18H], eax
       jmp      SHORT G_M5650_IG05
						;; size=58 bbWeight=0.50 PerfScore 15.75

G_M5650_IG04:  ;; offset=0060H
       mov      r8d, 50
       call     [System.Text.StringBuilder:AppendWithExpansion(byref,int):this]
						;; size=12 bbWeight=0.50 PerfScore 1.62

G_M5650_IG05:  ;; offset=006CH
       nop      
						;; size=1 bbWeight=1 PerfScore 0.25

G_M5650_IG06:  ;; offset=006DH
       add      rsp, 40
       ret      
						;; size=5 bbWeight=1 PerfScore 1.25
; Total bytes of code: 114
```

</details>

<details>
<summary>Append by index const</summary>

```csharp
public static void AppendConstIndex(StringBuilder builder)
{
    builder.Append("1234567890abcdefghijklmnopqrstuvwxyzABCDEFGHIJKLMN", 2, 16);
}
```

***Before***
```assembly
; Method Program:AppendConstIndex(System.Text.StringBuilder)
G_M5612_IG01:  ;; offset=0000H
       sub      rsp, 40
       vzeroupper 
						;; size=7 bbWeight=1 PerfScore 1.25

G_M5612_IG02:  ;; offset=0007H
       mov      rdx, 0x257002028D8      ; '1234567890abcdefghijklmnopqrstuvwxyzABCDEFGHIJKLMN'
       add      rdx, 12
       add      rdx, 4
       mov      r8, gword ptr [rcx+08H]
       mov      eax, dword ptr [rcx+18H]
       lea      r9d, [rax+10H]
       cmp      dword ptr [r8+08H], r9d
       jb       SHORT G_M5612_IG04
						;; size=35 bbWeight=1 PerfScore 9.25

G_M5612_IG03:  ;; offset=002AH
       movsxd   r9, eax
       lea      r8, bword ptr [r8+2*r9+10H]
       vmovdqu  ymm0, ymmword ptr [rdx]
       vmovdqu  ymmword ptr [r8], ymm0
       add      eax, 16
       mov      dword ptr [rcx+18H], eax
       jmp      SHORT G_M5612_IG05
						;; size=25 bbWeight=0.50 PerfScore 5.75

G_M5612_IG04:  ;; offset=0043H
       mov      r8d, 16
       call     [System.Text.StringBuilder:AppendWithExpansion(byref,int):this]
						;; size=12 bbWeight=0.50 PerfScore 1.62

G_M5612_IG05:  ;; offset=004FH
       nop      
						;; size=1 bbWeight=1 PerfScore 0.25

G_M5612_IG06:  ;; offset=0050H
       add      rsp, 40
       ret      
						;; size=5 bbWeight=1 PerfScore 1.25
; Total bytes of code: 85
```

***After***
```assembly
; Method Program:AppendConstIndex(System.Text.StringBuilder)
G_M5612_IG01:  ;; offset=0000H
       sub      rsp, 40
       vzeroupper 
						;; size=7 bbWeight=1 PerfScore 1.25

G_M5612_IG02:  ;; offset=0007H
       mov      rdx, 0x18E802028D8      ; '1234567890abcdefghijklmnopqrstuvwxyzABCDEFGHIJKLMN'
       add      rdx, 12
       add      rdx, 4
       mov      r8, gword ptr [rcx+08H]
       mov      eax, dword ptr [rcx+18H]
       lea      r9d, [rax+10H]
       cmp      dword ptr [r8+08H], r9d
       jb       SHORT G_M5612_IG04
						;; size=35 bbWeight=1 PerfScore 9.25

G_M5612_IG03:  ;; offset=002AH
       movsxd   r9, eax
       lea      r8, bword ptr [r8+2*r9+10H]
       vmovdqu  ymm0, ymmword ptr [rdx]
       vmovdqu  ymmword ptr [r8], ymm0
       add      eax, 16
       mov      dword ptr [rcx+18H], eax
       jmp      SHORT G_M5612_IG05
						;; size=25 bbWeight=0.50 PerfScore 5.75

G_M5612_IG04:  ;; offset=0043H
       mov      r8d, 16
       call     [System.Text.StringBuilder:AppendWithExpansion(byref,int):this]
						;; size=12 bbWeight=0.50 PerfScore 1.62

G_M5612_IG05:  ;; offset=004FH
       nop      
						;; size=1 bbWeight=1 PerfScore 0.25

G_M5612_IG06:  ;; offset=0050H
       add      rsp, 40
       ret      
						;; size=5 bbWeight=1 PerfScore 1.25
; Total bytes of code: 85
```
</details>

**Comment by EgorBo** [comment]
> I assume you will want to take it from here and close this?

Feel free to integrate into your PR so we can merge it

File: src/libraries/System.Private.CoreLib/src/System/Text/StringBuilder.cs [hunk 1]
```diff
@@ -728,7 +728,7 @@ public StringBuilder Append(char[]? value, int startIndex, int charCount)
         /// <param name="value">The string to append.</param>
         public StringBuilder Append(string? value)
         {
-            if (value is not null)
+            if (value != null)
```
Comment (EgorBo) [inline src/libraries/System.Private.CoreLib/src/System/Text/StringBuilder.cs:731]: I think you can revert this check, I was initially blaming Roslyn for extra stloc/ldloc but it turned out the reason was "reverted" args via prefixing that you fixed

**Review by EgorBo — APPROVED** [review]
Thanks for noticing the opportunity!

**Review by AndyAyersMS — APPROVED** [review]
LGTM

**Comment by EgorBo** [comment]
Performance improvements on arm64:

- https://github.com/dotnet/perf-autofiling-issues/issues/17665

**Comment by sebastienros** [comment]
This apparently is regressing Fortunes benchmarks with middleware in aspnet.

https://github.com/dotnet/runtime/issues/86113

**Comment by EgorBo** [comment]
> This apparently is regressing Fortunes benchmarks with middleware in aspnet.
> 
> #86113

Yes, this PR introduces more unrollings => more avx512 usages => throttle

This and other regressions are expected to be fixed by https://github.com/dotnet/runtime/pull/85551

**Comment by sebastienros** [comment]
Thanks! The one that keeps giving. Makes sense since I couldn't repro the regression on the newer machines (aspnet-perf-lin).

**Pull request merged**